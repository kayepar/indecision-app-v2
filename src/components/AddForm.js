import React, { useState } from 'react';
import { useOptionsContext } from '../context/optionsContext';

const AddForm = () => {
    const [error, setError] = useState(undefined);
    const { options, optionsDispatch } = useOptionsContext();

    const handleAddOption = (e) => {
        e.preventDefault();

        const option = e.target.elements.option.value.trim();
        const formError = validateForm(option);

        if (formError) {
            setError(formError);
        } else {
            optionsDispatch({ type: 'ADD_OPTION', option });
            e.target.elements.option.value = '';
            setError(undefined);
        }
    };

    const validateForm = (option) => {
        let error;

        if (!option) {
            error = 'Enter a valid option';
        } else if (ifOptionExists(option)) {
            error = `This option already exists (# ${getOptionIndex(option)})`;
        }

        return error;
    };

    const ifOptionExists = (option) => {
        return options.indexOf(option) > -1;
    };

    const getOptionIndex = (option) => {
        return options.findIndex((item) => item === option) + 1;
    };

    return (
        <div className="add-option-container">
            {error && <p className="add-option-error">{error}</p>}
            <form className="add-option" onSubmit={handleAddOption}>
                <input className="add-option__input" type="text" name="option" autoComplete="off" autoFocus></input>
                <button className="button">Add Option</button>
            </form>
        </div>
    );
};

export default AddForm;
