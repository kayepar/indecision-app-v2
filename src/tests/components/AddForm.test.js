import { render, screen, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import AddForm from '../../components/AddForm';
import OptionsProvider from '../../context/optionsContext';

afterEach(cleanup);

const renderComponents = () => {
    return render(
        <OptionsProvider>
            <AddForm />
        </OptionsProvider>
    );
};

const addOptions = (options) => {
    options.forEach((option) => {
        addOption(option);
    });
};

const addOption = (option) => {
    const { inputElement, buttonElement } = getElements();

    userEvent.type(inputElement, option);
    userEvent.click(buttonElement);
};

const getElements = () => {
    const inputElement = screen.getByRole('textbox');
    const buttonElement = screen.getByRole('button');

    return { inputElement, buttonElement };
};

describe('Tests for AddForm component', () => {
    describe('Initialization tests', () => {
        test('Should correctly render the AddForm component', () => {
            const { asFragment } = renderComponents();

            expect(asFragment()).toMatchSnapshot();
        });

        test('On load, should have focus on the input box', () => {
            const { getByRole } = renderComponents();
            const inputElement = getByRole('textbox');

            expect(inputElement).toHaveFocus();
        });

        test('Should be able to type into the input box', () => {
            const { getByRole } = renderComponents();
            const inputElement = getByRole('textbox');
            const optionText = 'React JS';

            userEvent.type(inputElement, optionText);

            expect(inputElement.value).toEqual(optionText);
        });
    });

    describe('Valid form submissions', () => {
        test('If input box has text and the button is clicked, should not show error message', () => {
            const { queryByText } = renderComponents();
            const { inputElement } = getElements();

            addOption('React JS');

            expect(inputElement.value).toEqual('');
            expect(queryByText(/Enter a valid option/i)).not.toBeInTheDocument();
        });

        test('If valid option is entered after an error, should clear error message', async () => {
            const { queryByText } = renderComponents();
            const { inputElement } = getElements();

            addOption(' '); // invalid option

            const errorElement = queryByText(/Enter a valid option/i);

            expect(errorElement).toBeInTheDocument();

            addOption('HTML 5'); // valid option

            expect(inputElement.value).toBe('');
            expect(errorElement).not.toBeInTheDocument();
        });

        test('If multiple unique options are added, should not show error message', () => {
            const options = ['React JS', 'Node JS', 'CSS'];
            const { queryByText } = renderComponents();

            addOptions(options);

            const errorElement = queryByText(/This option already exists/i);

            expect(errorElement).not.toBeInTheDocument();
        });

        test('Should submit form on enter key', () => {
            const { queryByText } = renderComponents();
            const { inputElement } = getElements();

            userEvent.type(inputElement, 'JAVA{enter}');

            const errorElement = queryByText(/Enter a valid option/i);

            expect(inputElement.value).toBe('');
            expect(errorElement).not.toBeInTheDocument();
        });
    });

    describe('Invalid form submissions', () => {
        test('If input box is empty and the button is clicked, should show error message', () => {
            const { getByText, getByRole } = renderComponents();

            userEvent.click(getByRole('button'));

            const errorElement = getByText(/Enter a valid option/i);

            expect(errorElement).toBeVisible();
        });

        test('If option already exist, should show error message', () => {
            const options = ['Javascript', 'Javascript'];
            const { queryByText } = renderComponents();

            addOptions(options);

            const errorElement = queryByText(/This option already exists \(#1\)/i);

            expect(errorElement).toBeInTheDocument();
        });

        test('If option already exist, text should remain in the input box', () => {
            const options = ['React JS', 'React JS'];
            const { queryByText } = renderComponents();
            addOptions(options);

            const { inputElement } = getElements();
            const errorElement = queryByText(/This option already exists/i);

            expect(errorElement).toBeInTheDocument();
            expect(inputElement.value).toBe(options[0]);
        });
    });
});
