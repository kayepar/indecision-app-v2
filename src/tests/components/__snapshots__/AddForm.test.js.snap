// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for AddForm component Initialization tests Should correctly render the AddForm component 1`] = `
<DocumentFragment>
  <div
    class="add-option-container"
  >
    <form
      class="add-option"
    >
      <input
        autocomplete="off"
        class="add-option__input"
        name="option"
        type="text"
      />
      <button
        class="button"
      >
        Add Option
      </button>
    </form>
  </div>
</DocumentFragment>
`;
